// @faustwp/core/dist/mjs/hooks/useAuth.js
import trim from 'lodash/trim.js';
import defaults from 'lodash/defaults.js';
import { useEffect, useState } from 'react';
import { gql } from '@apollo/client';
import { ensureAuthorization, } from '../auth/index.js';
import { getApolloAuthClient } from '../client.js';
export function useAuth(_config) {
    const config = defaults(_config, {
        strategy: 'redirect',
        shouldRedirect: false,
        skip: false,
    });
    if (config.strategy === 'local' && !config.loginPageUrl) {
        throw new Error('useAuth: Local strategies must specify the "loginPageUrl"');
    }
    const [isAuthenticated, setIsAuthenticated] = useState(null);
    const [isReady, setIsReady] = useState(false);
    const [loginUrl, setLoginUrl] = useState(null);
    const [called, setCalled] = useState(false);
    const [viewer, setViewer] = useState(null);
    useEffect(() => {
        if (config.skip === true) {
            return;
        }
        if (called) {
            return;
        }
        setCalled(true);
        const ensureAuthorizationConfig = {
            redirectUri: window.location.href,
        };
        if (config.strategy === 'local') {
            ensureAuthorizationConfig.loginPageUri = `/${trim(config.loginPageUrl, '/')}?redirect_uri=${encodeURIComponent(window.location.href)}`;
        }
        /* eslint-disable @typescript-eslint/no-floating-promises */
        (async () => {
            const authResult = await ensureAuthorization(ensureAuthorizationConfig);
            setIsAuthenticated(authResult === true);
            if (authResult !== true &&
                (authResult === null || authResult === void 0 ? void 0 : authResult.login) &&
                config.strategy === 'local') {
                setLoginUrl(authResult.login);
            }
            if (authResult !== true &&
                (authResult === null || authResult === void 0 ? void 0 : authResult.redirect) &&
                config.strategy === 'redirect') {
                setLoginUrl(authResult.redirect);
            }
            setIsReady(true);
        })();
    }, [called, config]);
    /**
     * Automatically redirect the user to the login page if the
     * shouldRedirect option is set to true.
     */
    useEffect(() => {
        if (config.skip === true) {
            return;
        }
        if (!config.shouldRedirect ||
            !isReady ||
            isAuthenticated !== false ||
            !loginUrl) {
            return;
        }
        /**
         * Using a setTimeout here because the page transition
         * is a little too fast and makes for bad UX.
         */
        setTimeout(() => {
            window.location.assign(loginUrl);
        }, 200);
    }, [isReady, isAuthenticated, loginUrl, config]);
    /**
     * Expose viewer options to the toolbar if the user is authenticated.
     */
    useEffect(() => {
        if (config.skip === true) {
            return;
        }
        if (isAuthenticated !== true) {
            return;
        }
        (async () => {
            const client = getApolloAuthClient();
            const { data } = await client.query({
                query: gql `
          query GetFaustViewer {
            viewer {
              name
              username
              capabilities
              databaseId
              description
              email
              firstName
              id
              lastName
              nickname
              locale
              registeredDate
              slug
              templates
              uri
              url
              userId
            }
          }
        `,
            });
            setViewer(data.viewer);
        })();
    }, [isAuthenticated, config.skip]);
    return {
        isAuthenticated,
        isReady,
        loginUrl,
        viewer,
    };
}
