// @faustwp/core/dist/mjs/index.js
import { FaustProvider } from './components/FaustProvider.js';
import { WordPressTemplate, } from './components/WordPressTemplate.js';
import { getWordPressProps } from './getWordPressProps.js';
import { getNextStaticProps, getNextServerSideProps, } from './getProps.js';
import { getConfig, setConfig } from './config/index.js';
import { ensureAuthorization } from './auth/index.js';
import { authorizeHandler, logoutHandler, apiRouter } from './server/index.js';
import { withFaust } from './config/withFaust.js';
import { getWpUrl } from './lib/getWpUrl.js';
import { getAdminUrl } from './lib/getAdminUrl.js';
import { getGraphqlEndpoint } from './lib/getGraphqlEndpoint.js';
import { getWpHostname } from './lib/getWpHostname.js';
import { getApolloClient, getApolloAuthClient, addApolloState, } from './client.js';
import { hooks } from './wpHooks/index.js';
import { getSitemapProps, } from './server/sitemaps/getSitemapProps.js';
import { useAuth } from './hooks/useAuth.js';
import { useLogin } from './hooks/useLogin.js';
import { useLogout } from './hooks/useLogout.js';
import { useFaustQuery } from './hooks/useFaustContext.js';
import { FaustContext } from './store/FaustContext.js';
import { ToolbarNode, ToolbarItem, ToolbarNodeSkeleton, ToolbarSubmenu, ToolbarSubmenuWrapper, } from './components/Toolbar/index.js';
import { flatListToHierarchical } from './utils/flatListToHierarchical.js';
export { flatListToHierarchical, FaustProvider, WordPressTemplate, getWordPressProps, getNextStaticProps, getNextServerSideProps, getConfig, setConfig, ensureAuthorization, authorizeHandler, logoutHandler, apiRouter, withFaust, getWpHostname, getWpUrl, getAdminUrl, getGraphqlEndpoint, getApolloClient, getApolloAuthClient, addApolloState, getSitemapProps, useAuth, useLogin, useLogout, ToolbarNode, ToolbarItem, ToolbarNodeSkeleton, ToolbarSubmenu, ToolbarSubmenuWrapper, hooks, useFaustQuery, FaustContext, };
