// @faustwp/blocks/dist/mjs/blocks/CoreHeading.js
import { gql } from '@apollo/client';
import React from 'react';
import { useBlocksTheme } from '../components/WordPressBlocksProvider.js';
import { getStyles } from '../utils/index.js';
export function CoreHeading(props) {
    var _a;
    const theme = useBlocksTheme();
    const style = getStyles(theme, Object.assign({}, props));
    const { attributes } = props;
    const level = `h${(_a = attributes === null || attributes === void 0 ? void 0 : attributes.level) !== null && _a !== void 0 ? _a : 1}`;
    const headingProps = {
        style,
        dangerouslySetInnerHTML: { __html: attributes === null || attributes === void 0 ? void 0 : attributes.content },
        id: attributes === null || attributes === void 0 ? void 0 : attributes.anchor,
        className: attributes === null || attributes === void 0 ? void 0 : attributes.cssClassName,
    };
    return React.createElement(level, headingProps);
}
CoreHeading.fragments = {
    key: `CoreHeadingBlockFragment`,
    entry: gql `
    fragment CoreHeadingBlockFragment on CoreHeading {
      attributes {
        align
        anchor
        backgroundColor
        content
        fontFamily
        fontSize
        gradient
        level
        style
        textAlign
        textColor
        cssClassName
      }
    }
  `,
};
CoreHeading.config = {
    name: 'CoreHeading',
};
CoreHeading.displayName = 'CoreHeading';
