// @faustwp/core/dist/cjs/server/sitemaps/getSitemapProps.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSitemapProps = exports.validateConfig = exports.FAUST_PAGES_PATHNAME = exports.SITEMAP_INDEX_PATH = void 0;
const isUndefined_js_1 = __importDefault(require("lodash/isUndefined.js"));
const isString_js_1 = __importDefault(require("lodash/isString.js"));
const isArray_js_1 = __importDefault(require("lodash/isArray.js"));
const isObject_js_1 = __importDefault(require("lodash/isObject.js"));
const convert_js_1 = require("../../utils/convert.js");
const createSitemaps_js_1 = require("./createSitemaps.js");
/**
 * The pathname to the root sitemap index.
 */
exports.SITEMAP_INDEX_PATH = '/sitemap.xml';
/**
 * The pathname to the Next.js pages sitemap file. We may want to make this
 * configurable in the future.
 */
exports.FAUST_PAGES_PATHNAME = '/sitemap-faust-pages.xml';
/**
 * Validates the structure of the user defined config.
 *
 * @param {Partial<HandleSitemapRequestsConfig>} config The user provided config
 */
function validateConfig(config) {
    var _a, _b;
    if (!(0, isString_js_1.default)(config === null || config === void 0 ? void 0 : config.frontendUrl)) {
        throw new Error('frontendUrl must be a string');
    }
    try {
        // eslint-disable-next-line no-new
        const url = new URL(config.frontendUrl);
        if (url.protocol !== 'http:' && url.protocol !== 'https:') {
            throw new Error('URL must have protocol');
        }
    }
    catch (e) {
        throw new Error('frontendUrl must be a valid URL.');
    }
    if (!(0, isUndefined_js_1.default)(config === null || config === void 0 ? void 0 : config.sitemapPathsToIgnore)) {
        if (!(0, isArray_js_1.default)(config.sitemapPathsToIgnore)) {
            throw new Error('sitemapPathsToIgnore must be an array');
        }
        (_a = config === null || config === void 0 ? void 0 : config.sitemapPathsToIgnore) === null || _a === void 0 ? void 0 : _a.forEach((path) => {
            if (!(0, isString_js_1.default)(path)) {
                throw new Error('sitemapPathsToIgnore must be an array of strings');
            }
            if (!path.startsWith('/')) {
                throw new Error('Each sitemapPathsToIgnore must start with a forward slash');
            }
            if (path.includes('*') && !path.endsWith('*')) {
                throw new Error('sitemapPathsToIgnore with a wildcard must end with a wildcard');
            }
        });
    }
    // Validate pages structure and required values
    if (!(0, isUndefined_js_1.default)(config === null || config === void 0 ? void 0 : config.pages)) {
        if (!(0, isArray_js_1.default)(config.pages)) {
            throw new Error('pages must be an array');
        }
        (_b = config === null || config === void 0 ? void 0 : config.pages) === null || _b === void 0 ? void 0 : _b.forEach((page) => {
            if (!(0, isObject_js_1.default)(page)) {
                throw new Error('pages must be an array of objects');
            }
            if ((0, isUndefined_js_1.default)(page.path)) {
                throw new Error('pages must have a path property');
            }
            if (!(0, isString_js_1.default)(page.path)) {
                throw new Error('The pages path property must be a string');
            }
        });
    }
}
exports.validateConfig = validateConfig;
async function getSitemapProps(ctx, config) {
    // config validation with middleware flag
    validateConfig(config);
    if (!ctx.req.url) {
        throw new Error('A context url is required.');
    }
    const queryParam = (0, convert_js_1.getQueryParam)(ctx.req.url, 'sitemap');
    if (queryParam === '') {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        const response = await (0, createSitemaps_js_1.createRootSitemapIndex)(ctx.req, config);
        if (!response || (response === null || response === void 0 ? void 0 : response.status) === 404) {
            return {
                notFound: true,
            };
        }
        ctx.res.setHeader('Content-Type', 'application/xml');
        ctx.res.write(await (response === null || response === void 0 ? void 0 : response.text()));
        ctx.res.end();
    }
    if (queryParam !== '' && queryParam !== 'sitemap-faust-pages.xml') {
        const response = await (0, createSitemaps_js_1.handleSitemapPath)(ctx.req, config);
        if (!response || (response === null || response === void 0 ? void 0 : response.status) === 404) {
            return {
                notFound: true,
            };
        }
        ctx.res.setHeader('Content-Type', 'application/xml');
        ctx.res.write(await (response === null || response === void 0 ? void 0 : response.text()));
        ctx.res.end();
    }
    if (queryParam !== '' && queryParam === 'sitemap-faust-pages.xml') {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
        const response = (0, createSitemaps_js_1.createPagesSitemap)(ctx.req, config);
        ctx.res.setHeader('Content-Type', 'application/xml');
        ctx.res.write(await (response === null || response === void 0 ? void 0 : response.text()));
        ctx.res.end();
    }
    return {
        props: {},
    };
}
exports.getSitemapProps = getSitemapProps;
