// @faustwp/core/dist/cjs/hooks/useAuth.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useAuth = void 0;
const trim_js_1 = __importDefault(require("lodash/trim.js"));
const defaults_js_1 = __importDefault(require("lodash/defaults.js"));
const react_1 = require("react");
const client_1 = require("@apollo/client");
const index_js_1 = require("../auth/index.js");
const client_js_1 = require("../client.js");
function useAuth(_config) {
    const config = (0, defaults_js_1.default)(_config, {
        strategy: 'redirect',
        shouldRedirect: false,
        skip: false,
    });
    if (config.strategy === 'local' && !config.loginPageUrl) {
        throw new Error('useAuth: Local strategies must specify the "loginPageUrl"');
    }
    const [isAuthenticated, setIsAuthenticated] = (0, react_1.useState)(null);
    const [isReady, setIsReady] = (0, react_1.useState)(false);
    const [loginUrl, setLoginUrl] = (0, react_1.useState)(null);
    const [called, setCalled] = (0, react_1.useState)(false);
    const [viewer, setViewer] = (0, react_1.useState)(null);
    (0, react_1.useEffect)(() => {
        if (config.skip === true) {
            return;
        }
        if (called) {
            return;
        }
        setCalled(true);
        const ensureAuthorizationConfig = {
            redirectUri: window.location.href,
        };
        if (config.strategy === 'local') {
            ensureAuthorizationConfig.loginPageUri = `/${(0, trim_js_1.default)(config.loginPageUrl, '/')}?redirect_uri=${encodeURIComponent(window.location.href)}`;
        }
        /* eslint-disable @typescript-eslint/no-floating-promises */
        (async () => {
            const authResult = await (0, index_js_1.ensureAuthorization)(ensureAuthorizationConfig);
            setIsAuthenticated(authResult === true);
            if (authResult !== true &&
                (authResult === null || authResult === void 0 ? void 0 : authResult.login) &&
                config.strategy === 'local') {
                setLoginUrl(authResult.login);
            }
            if (authResult !== true &&
                (authResult === null || authResult === void 0 ? void 0 : authResult.redirect) &&
                config.strategy === 'redirect') {
                setLoginUrl(authResult.redirect);
            }
            setIsReady(true);
        })();
    }, [called, config]);
    /**
     * Automatically redirect the user to the login page if the
     * shouldRedirect option is set to true.
     */
    (0, react_1.useEffect)(() => {
        if (config.skip === true) {
            return;
        }
        if (!config.shouldRedirect ||
            !isReady ||
            isAuthenticated !== false ||
            !loginUrl) {
            return;
        }
        /**
         * Using a setTimeout here because the page transition
         * is a little too fast and makes for bad UX.
         */
        setTimeout(() => {
            window.location.assign(loginUrl);
        }, 200);
    }, [isReady, isAuthenticated, loginUrl, config]);
    /**
     * Expose viewer options to the toolbar if the user is authenticated.
     */
    (0, react_1.useEffect)(() => {
        if (config.skip === true) {
            return;
        }
        if (isAuthenticated !== true) {
            return;
        }
        (async () => {
            const client = (0, client_js_1.getApolloAuthClient)();
            const { data } = await client.query({
                query: (0, client_1.gql) `
          query GetFaustViewer {
            viewer {
              name
              username
              capabilities
              databaseId
              description
              email
              firstName
              id
              lastName
              nickname
              locale
              registeredDate
              slug
              templates
              uri
              url
              userId
            }
          }
        `,
            });
            setViewer(data.viewer);
        })();
    }, [isAuthenticated, config.skip]);
    return {
        isAuthenticated,
        isReady,
        loginUrl,
        viewer,
    };
}
exports.useAuth = useAuth;
