// @faustwp/blocks/dist/cjs/blocks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable object-shorthand */
const CoreParagraph_js_1 = require("./CoreParagraph.js");
const CoreColumns_js_1 = require("./CoreColumns.js");
const CoreColumn_js_1 = require("./CoreColumn.js");
const CoreButtons_js_1 = require("./CoreButtons.js");
const CoreButton_js_1 = require("./CoreButton.js");
const CoreCode_js_1 = require("./CoreCode.js");
const CoreQuote_js_1 = require("./CoreQuote.js");
const CoreImage_js_1 = require("./CoreImage.js");
const CoreSeparator_js_1 = require("./CoreSeparator.js");
const CoreList_js_1 = require("./CoreList.js");
const CoreHeading_js_1 = require("./CoreHeading.js");
exports.default = {
    CoreParagraph: CoreParagraph_js_1.CoreParagraph,
    CoreColumns: CoreColumns_js_1.CoreColumns,
    CoreColumn: CoreColumn_js_1.CoreColumn,
    CoreCode: CoreCode_js_1.CoreCode,
    CoreQuote: CoreQuote_js_1.CoreQuote,
    CoreImage: CoreImage_js_1.CoreImage,
    CoreSeparator: CoreSeparator_js_1.CoreSeparator,
    CoreList: CoreList_js_1.CoreList,
    CoreButton: CoreButton_js_1.CoreButton,
    CoreButtons: CoreButtons_js_1.CoreButtons,
    CoreHeading: CoreHeading_js_1.CoreHeading,
};
